vagrant shell provisioning errors

init_environment.sh has a couple of annoyances and some errors:

I do not know how to resolve these, I will try to install the latest version for ember-cli, but I do not know how it will affect your project.
* npm throws "unmet dependency" warnings about debug when installing ember-cli (upgrading ember-cli could help according to https://github.com/npm/npm/issues/8241)

Example:
```
==> default: npm
==> default:
==> default: WARN
==> default:
==> default: unmet dependency
==> default:  /usr/lib/node_modules/ember-cli/node_modules/express requires debug@'~2.1.1' but will load
```
Full log: https://gist.github.com/1cbfca3c2a8e1677000c.git

* ember throws an error when building
```
==> default: version: 0.1.7
==> default: Could not find watchman, falling back to NodeWatcher for file system events
==> default: Building
==> default: Building...
==> default: Cannot find module 'ember-cli/lib/broccoli/ember-app'
==> default: Error: Cannot find module 'ember-cli/lib/broccoli/ember-app'
==> default:     at Function.Module._resolveFilename (module.js:338:15)
==> default:     at Function.Module._load (module.js:280:25)
==> default:     at Module.require (module.js:364:17)
==> default:     at require (module.js:380:17)
==> default:     at Object.<anonymous> (/var/www/twweb/Brocfile.js:3:16)
==> default:     at Module._compile (module.js:456:26)
==> default:     at Object.Module._extensions..js (module.js:474:10)
==> default:     at Module.load (module.js:356:32)
==> default:     at Function.Module._load (module.js:312:12)
==> default:     at Module.require (module.js:364:17)
==> default:     at require (module.js:380:17)
```

* python managepy run throws error:
```
(twweb)vagrant@precise32:/var/www/twweb$ python manage.py run
 Note: It will take a few seconds for both necessary
      servers to start.  Once you see the message
      message 'Build Successful', both servers are up.

Validating models...

version: 0.1.7
0 errors found
October 08, 2015 - 17:54:29
Django version 1.6.1, using settings 'inthe_am.settings'
Starting development server at http://0.0.0.0:8001/
Quit the server with CONTROL-C.
Could not find watchman, falling back to NodeWatcher for file system events
Cannot find module '/var/www/twweb/node_modules/broccoli-ember-hbs-template-compiler'
Error: Cannot find module '/var/www/twweb/node_modules/broccoli-ember-hbs-template-compiler'
    at Function.Module._resolveFilename (module.js:338:15)
    at Function.Module._load (module.js:280:25)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at requireLocal (/usr/lib/node_modules/ember-cli/lib/utilities/require-local.js:6:10)
    at TemplatePlugin.toTree (/usr/lib/node_modules/ember-cli/lib/preprocessors/template-plugin.js:16:10)
    at /usr/lib/node_modules/ember-cli/lib/preprocessors.js:104:26
    at Array.forEach (native)
    at processPlugins (/usr/lib/node_modules/ember-cli/lib/preprocessors.js:103:11)
    at module.exports.preprocessTemplates (/usr/lib/node_modules/ember-cli/lib/preprocessors.js:86:10)
```

* npm throws "No README data" warnings
Example:
```
==> default: npm
==> default:
==> default: WARN
==> default:
==> default: package.json
==> default:  broccoli-asset-rev@2.2.0 No README data
```

Full log: https://gist.github.com/1cbfca3c2a8e1677000c.git

I have attached a patch for these following two annoyances, I did not want to create a pull request for something so small:
* in inthe_am/settings.py line 65, MIDDLEWARE_CLASSES should be list, not tuple
* turn off wget progress bar

Patch: https://gist.github.com/222a0dc6cbfcaed52f28.git

Which version of npm, ember and bower you are using? Can you attach npm list output?

The gists were not private, but I did not embed the URL's right, here is the correct one:

{% gist 1cbfca3c2a8e1677000c &}

I am still getting the following error when running, this prevents the second server (8000) from running:

```
Could not find watchman, falling back to NodeWatcher for file system events
Cannot find module '/var/www/twweb/node_modules/broccoli-ember-hbs-template-compiler'
Error: Cannot find module '/var/www/twweb/node_modules/broccoli-ember-hbs-template-compiler'
    at Function.Module._resolveFilename (module.js:338:15)
    at Function.Module._load (module.js:280:25)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at requireLocal (/usr/lib/node_modules/ember-cli/lib/utilities/require-local.js:6:10)
    at TemplatePlugin.toTree (/usr/lib/node_modules/ember-cli/lib/preprocessors/template-plugin.js:16:10)
    at /usr/lib/node_modules/ember-cli/lib/preprocessors.js:104:26
    at Array.forEach (native)
    at processPlugins (/usr/lib/node_modules/ember-cli/lib/preprocessors.js:103:11)
    at module.exports.preprocessTemplates (/usr/lib/node_modules/ember-cli/lib/preprocessors.js:86:10)
```

Upgrading to latest version of ember and bower seems to fix the issue. If that is ok with you, I can send a seperate pull request for that too.

I will send a pull request for the previous patch.

--- a/inthe_am/settings.py
+++ b/inthe_am/settings.py
@@ -62,14 +62,14 @@ INSTALLED_APPS = (
     'silk',
 )
 
-MIDDLEWARE_CLASSES = (
+MIDDLEWARE_CLASSES = [
     'django.contrib.sessions.middleware.SessionMiddleware',
     'django.middleware.common.CommonMiddleware',
     'django.middleware.csrf.CsrfViewMiddleware',
     'django.contrib.auth.middleware.AuthenticationMiddleware',
     'django.contrib.messages.middleware.MessageMiddleware',
     'django.middleware.clickjacking.XFrameOptionsMiddleware',
-)
+]
 
 AUTHENTICATION_BACKENDS = (
     'social.backends.google.GoogleOAuth2',

--- a/scripts/vagrant/init_environment.sh
+++ b/scripts/vagrant/init_environment.sh
@@ -112,13 +112,28 @@ if [ $? -ne 0 ]; then
     cp /var/taskd/src/task-2.3.0/task_2.3.0-1*.deb /tmp
 fi
 
-cd $MAIN_DIR
-npm install -g ember-cli@0.1.7 bower@1.3.12
-ember install
-npm install
-bower --config.interactive=false install --allow-root
+# copy MAIN_DIR into a temp folder and run npm there
+# to avoid EPERM errors on NFS shared folders in vagrant
+# See: https://github.com/npm/npm/issues/3565
+cd $STARTING_DIR
+cp -a $MAIN_DIR /tmp/twweb
+chown -R vagrant. /tmp/twweb
+cd /tmp/twweb
 
+echo "installing ember-cli and bower"
+npm install -g ember-cli bower
+echo "running npm install"
+npm install
+echo "running bower install"
+bower --config.interactive=false install --allow-root
+echo "running ember install"
+ember install
+echo "running ember build"
 ember build
+
+# Sync back node_modules to original place
+rsync --recursive --links --times /tmp/twweb/ $MAIN_DIR/
+cd $MAIN_DIR
 
 # Install requirements
 source /var/www/envs/twweb/bin/activate

--- a/scripts/vagrant/init_environment.sh
+++ b/scripts/vagrant/init_environment.sh
@@ -1,3 +1,6 @@
+#!/bin/bash
+set -e
+
 STARTING_DIR=$(pwd)
 MAIN_DIR=/var/www/twweb
 if [ ! -z "$TRAVIS" ]; then
@@ -5,6 +8,7 @@ if [ ! -z "$TRAVIS" ]; then
 fi
 
 # Install necessary packages
+echo "installing depencencies"
 apt-get update
 apt-get install -y python-software-properties
 apt-add-repository -y ppa:chris-lea/node.js
@@ -13,18 +17,20 @@ apt-get install -y git postgresql-server-dev-9.1 python-dev cmake build-essentia
 
 if [ -z "$TRAVIS" ]; then
     PHANTOMJS=phantomjs-1.9.7-linux-i686
+    echo "installing $PHANTOMJS"
     cd /usr/local/share/
     if [ ! -d $PHANTOMJS ]; then
-        wget https://bitbucket.org/ariya/phantomjs/downloads/$PHANTOMJS.tar.bz2
+        wget -nv https://bitbucket.org/ariya/phantomjs/downloads/$PHANTOMJS.tar.bz2
         tar -xjf $PHANTOMJS.tar.bz2
         ln -s /usr/local/share/$PHANTOMJS/bin/phantomjs /usr/local/share/phantomjs; sudo ln -s /usr/local/share/$PHANTOMJS/bin/phantomjs /usr/local/bin/phantomjs; sudo ln -s /usr/local/share/$PHANTOMJS/bin/phantomjs /usr/bin/phantomjs
     fi
     cd $STARTING_DIR
 
     # Set up virtual environment
+    echo "setting up virtualenv"
     mkdir -p /var/www/envs
     if [ ! -d /var/www/envs/twweb ]; then
-        wget https://raw.github.com/pypa/pip/master/contrib/get-pip.py
+        wget -nv https://raw.github.com/pypa/pip/master/contrib/get-pip.py
         python get-pip.py
         pip install virtualenv
         virtualenv /var/www/envs/twweb
@@ -52,21 +58,25 @@ mkdir -p $MAIN_DIR/logs
 if [ ! -d $TWWEB_TASKD_DATA ]; then
     # See environment variable TWWEB_TASKD_DATA
 
+    TASKD_VERSION="taskd-1.0.0"
+    echo "installing $TASKD_VERSION and setup certificates"
     mkdir -p $TWWEB_TASKD_DATA/src
+    sudo chown -R vagrant. $TWWEB_TASKD_DATA
     cd $TWWEB_TASKD_DATA/src
 
-    wget http://taskwarrior.org/download/taskd-1.0.0.tar.gz
-    tar xzf taskd-1.0.0.tar.gz
-    cd taskd-1.0.0
+    wget -nv http://taskwarrior.org/download/$TASKD_VERSION.tar.gz
+    tar xzf $TASKD_VERSION.tar.gz
+    cd $TASKD_VERSION
 
-    which taskd
-    if [ $? -ne 0 ]; then
+    set +e
+    which taskd; RETVAL=$?
+    set -e
+    if [ $RETVAL -ne 0 ]; then
         cmake .
         make
         checkinstall --default
-        cp  /var/taskd/src/taskd-1.0.0/taskd_1.0.0-1*.deb /tmp
+        cp  /var/taskd/src/$TASKD_VERSION/*.deb /tmp
     fi
-
 
     cd $TWWEB_TASKD_DATA
     export TASKDDATA=$TWWEB_TASKD_DATA
@@ -75,12 +85,12 @@ if [ ! -d $TWWEB_TASKD_DATA ]; then
     taskd add org testing
     cp $MAIN_DIR/scripts/vagrant/simple_taskd_upstart.conf /etc/init/taskd.conf
 
-    if [ -z "$TRAVIS" ]; then
+    if [ -z "$TRAVIS" && -z "`pgrep -x taskd`" ]; then
         service taskd stop
     fi
 
     # generate certificates
-    cd $TWWEB_TASKD_DATA/src/taskd-1.0.0/pki
+    cd $TWWEB_TASKD_DATA/src/$TASKD_VERSION/pki
     ./generate
     cp client.cert.pem $TASKDDATA
     cp client.key.pem $TASKDDATA
@@ -98,18 +108,22 @@ if [ ! -d $TWWEB_TASKD_DATA ]; then
     if [ -z "$TRAVIS" ]; then
         service taskd start
     fi
-fi
 
-which task
-if [ $? -ne 0 ]; then
-    cd $TWWEB_TASKD_DATA/src
-    wget http://taskwarrior.org/download/task-2.3.0.tar.gz
-    tar xzf task-2.3.0.tar.gz
-    cd task-2.3.0
-    cmake .
-    make
-    checkinstall --default
-    cp /var/taskd/src/task-2.3.0/task_2.3.0-1*.deb /tmp
+    set +e
+    which task; RETVAL=$?
+    set -e
+    if [ $RETVAL -ne 0 ]; then
+        TASK_VERSION="task-2.3.0"
+        echo "installing $TASK_VERSION"
+        cd $TWWEB_TASKD_DATA/src
+        wget -nv http://taskwarrior.org/download/$TASK_VERSION.tar.gz
+        tar xzf $TASK_VERSION.tar.gz
+        cd $TASK_VERSION
+        cmake .
+        make
+        checkinstall --default
+        cp /var/taskd/src/$TASK_VERSION/*.deb /tmp
+    fi
 fi
 
 # copy MAIN_DIR into a temp folder and run npm there
@@ -136,10 +150,12 @@ rsync --recursive --links --times /tmp/twweb/ $MAIN_DIR/
 cd $MAIN_DIR
 
 # Install requirements
+echo "installing python requirements"
 source /var/www/envs/twweb/bin/activate
 pip install --download-cache=/tmp/pip_cache -r $MAIN_DIR/requirements.txt
 
 if [ -z "$TRAVIS" ]; then
+    echo "preparing application"
     pip install ipdb
     python $MAIN_DIR/manage.py syncdb --noinput
     python $MAIN_DIR/manage.py migrate --noinput

